tree/treeplayer/src/TSimpleAnalysis.cxx

```diff
@@ -0,0 +1,286 @@
+ *************************************************************************/
+
+
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <vector>
+
+#include "TSimpleAnalysis.h"
+#include "TFile.h"
+#include "TCanvas.h"
+#include "TChain.h"
+#include "TH1.h"
+#include "TError.h"
+
+/** \class TSympleAnalysis
+A TSimpleAnalysis object permit to, given an input file or through
+command line, create an .root file in which are saved histograms
+that the user want to create.
+*/
+
+const std::string TSimpleAnalysis::kCutIntr="if";
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Constructor for the case with the input file
+///
+/// \param[in] kFile name of the input file that has to be read
+
+TSimpleAnalysis::TSimpleAnalysis(const std::string& kFile)
+{
+   fInputName = kFile;
+   in.open(kFile);
+   if(!in) {
+      ::Error("TSimpleAnalysis","File %s not found",kFile.c_str());
+      throw 1;
+   }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Constructor for the case of command line parsing arguments
+///
+/// \param[in] output name of the output file
+/// \param[in] inputFiles name of the input .root files
+/// \param[in] name name of the tree
+/// \param[in] expressions what is showed in the histograms
+
+TSimpleAnalysis::TSimpleAnalysis(const std::string& output, std::vector<std::string> inputFiles,
+                                 const std::string& name, std::vector<std::string> expressions)
+{
+   fOutputFile=output;
+   fTreeName=name;
+   for (std::string input: inputFiles)
+      fInputFiles.push_back(input);
+   for (std::string expr: expressions) {
+      std::size_t equal=expr.find("=");
+      if (equal == std::string::npos) {
+         ::Error("TSimpleAnalysis",
+                 "Missing '=' in fExpressions in %s",expr.c_str());
+         throw 1;
+      }
+      std::size_t cutPos=expr.find(kCutIntr, equal);
+      std::string substring=expr.substr(0,equal);
+      if (substring.empty()) {
+         ::Error("TSimpleAnalysis",
+                 "No hname found in %s",expr.c_str());
+         throw 2;
+      }
+      fHNames.push_back(substring);
+      substring=expr.substr(equal+1,cutPos-equal-1);
+      if (substring.empty()) {
+         ::Error("TSimpleAnalysis",
+                 "No expression found in %s",expr.c_str());
+         throw 3;
+      }
+      fExpressions.push_back(substring);
+      if (cutPos == std::string::npos) {
+         fCut.push_back("");
+      } else {
+         fCut.push_back(expr.substr(cutPos+kCutIntr.size()));
+      }
+   }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// Tell if the name of some histograms that will be created are the same
+///
+/// param[in] name name of the histograms
+
+void TSimpleAnalysis::CheckHNames(std::vector<std::string> name)
+{
+   Int_t err=0;
+   for (unsigned i=0; i<name.size()-1; i++)
```

Comment:
There is an algorithm for that. The trick is to call `std::sort(name)` first; then `std::adjacent_find(name.begin(), name.end())` will return `name.end()` if the names are all unique, and an iterator to the first of the duplicate names otherwise. You can repeat this (using the returned iterator + 2) to collect all dupes. See e.g. http://en.cppreference.com/w/cpp/algorithm/adjacent_find

The advantages: less code, guaranteed no off-by-one errors.